src/mixins/createI18nMixin.ts

```diff
@@ -0,0 +1,362 @@
+	 */
+	localizeBundle(bundle: Bundle<T>): T;
+
+	/**
+	 * Register a bundle with the instance so that `state.labels` can be mapped to labels on the bundle.
+	 *
+	 * @param bundle
+	 * The required bundle object to register with the instance.
+	 */
+	registerBundle(bundle: Bundle<T>): void;
+}
+
+export interface I18nFactory extends ComposeFactory<I18n<Messages>, I18nOptions<I18nState>> {}
+
+/**
+ * A map of `state` properties to bundle:label pairs.
+ */
+export interface I18nLabels {
+	[key: string]: string;
+}
+
+export interface I18nOptions<S extends I18nState> extends StatefulOptions<S> {
+	/**
+	 * An optional list of bundle objects to be registered with the instance.
+	 */
+	bundles?: Bundle<Messages>[];
+}
+
+export interface I18nState extends State {
+	/**
+	 * A map of `state` properties to bundle:label pairs.
+	 *
+	 * Each key on the object represents a state property that is set when the locale-specific messages are loaded.
+	 * If more than one bundle has been registered for the instnace, then each value must contain both the bundle
+	 * path and the label key, separated by a colon. For example, if the bundle path is "nls/greetings", and the key
+	 * is "goodbye", then the value is formatted as "nls/greetings:goodbye". However, if only one bundle has been
+	 * registered, then the label value can consist of just the label key (e.g., "goodbye").
+	 * Example:
+	 *
+	 * const widget = createWidget({
+	 *   bundles: [ greetingsBundle, otherBundle ],
+	 *   locale: 'es',
+	 *   state: {
+	 *     labels: {
+	 *       greeting: 'nls/greetings:goodDay
+	 *    }
+	 *  }
+	 * });
+	 *
+	 * In this example, once the i18n system has loaded the messages, the widget state is updated with a `greeting`
+	 * property whose value is correct locale-specific message (e.g., 'Buenos Dias').
+	 */
+	labels?: I18nLabels;
+
+	/**
+	 * The locale for the widget. Is not specified, then the next available parent locale is assumed (all the way up
+	 * to the root locale as determined by `dojo-I18n`). If specified, the widget's node will have a `data-locale`
+	 * property set to the locale, in order to facilitate styling localized components if the use case arises.
+	 */
+	locale?: string;
+
+	/**
+	 * An optional flag indicating the widget's text direction. If `true`, then the underlying node's `dir`
+	 * property is set to "rtl". If it is `false`, then the `dir` property is set to "ltr". Otherwise, the property
+	 * is not set.
+	 */
+	rtl?: boolean;
+}
+
+/**
+ * @private
+ * An internal helper interface for defining locale and text direction attributes on widget nodes.
+ */
+interface I18nVNodeProperties extends VNodeProperties {
+	'data-locale': string | null;
+	dir: string | null;
+}
+
+export type I18nMixin<M extends Messages, S extends I18nState> = I18n<M> & Stateful<S>;
+
+/**
+ * A map of bundles registered to each instance.
+ */
+const bundleMap = new WeakMap<I18nMixin<Messages, I18nState>, BundleMap<Messages>>();
```

Comment:
Ideally, since the key is the same for all these WeakMaps, it would be better to create an interface for all of them and store all this private data in the same weak map.